pkg/tenantctl/controller.go

```diff
@@ -157,219 +167,394 @@ func (c *Controller) processNextWorkItem() bool {
 // converge the two. It then updates the Status block of the tenant resource
 // with the current status of the resource.
 func (c *Controller) syncHandler(key string) error {
-	tenant, err := c.fetcher.ByKey(key)
+	origTenant, err := c.fetcher.ByKey(key)
 	if err != nil {
 		return err
 	}
 
-	if tenant == nil {
+	if origTenant == nil {
 		return nil
 	}
 
-	// Check if object has deletion timestamp
-	// If not, try to add finalizer if missing
-	if tenant.ObjectMeta.DeletionTimestamp.IsZero() {
-		changed, finalizerList := utils.AddStringIfMissing(tenant.ObjectMeta.Finalizers, k8s.FinalizerName)
-		if changed {
-			tenant.ObjectMeta.Finalizers = finalizerList
-			_, err = c.update(tenant)
-			return err
+	tenant := origTenant.DeepCopy()
+
+	c.logPrintln(tenant, "started reconciliation")
+	defer c.logPrintln(&api.Tenant{ObjectMeta: *tenant.ObjectMeta.DeepCopy()}, "finished reconciliation")
+
+	// the configuration should be loaded once per sync to avoid inconsistencies
+	// in case of concurrent configuration changes
+	config, err := getClientConfig(c.factory, tenant.GetNamespace())
+	if err != nil {
+		c.logPrintln(tenant, err)
+		return err
+	}
+
+	if !tenant.ObjectMeta.DeletionTimestamp.IsZero() {
+		c.logPrintln(tenant, "tenant is marked as deleted")
+		if !c.hasFinalizer(tenant) {
```

Comment:
We should try to delete the tenant namespace regardless finalizer is set or not.

Reply:
The controller implementation guarantees that:

-  before any dependent namespace is created, our finalizer has been added successfully.
-  the finalizer will be removed only after the tenant namespace has been successfully removed.

Therefore, if our finalizer is not present anymore, the dependent namespace either has been successfully cleaned up already or it has never been created. Repeating the namespace deletion is not necessary.

If an administrator manually removes our finalizer, he/she has to accept that the clean-up does not happen and an orphaned tenant namespace is left behind.

Also, once the tenant namespace has been deleted, its name is free to be used for other purposes. If the tenant resource would continue to delete this namespace, it potentially destroys productive data. However, there's still a chance for such case to happen: The rollback was executed but the finalizer removal failed. But the probability is much lower that repeating the namespace deletion until the tenant resource gets removed completely.